libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/Messages.scala

```diff
@@ -38,11 +37,33 @@ object MessageCodecs {
     } yield DeviceEventMessage(ns, event)
   }
 
-  implicit val deviceUpdateStatusEncoder: Encoder[DeviceUpdateStatus] = Encoder.enumEncoder(DeviceUpdateStatus)
-  implicit val deviceUpdateStatusDecoder: Decoder[DeviceUpdateStatus] = Decoder.enumDecoder(DeviceUpdateStatus)
+  implicit val deviceUpdateEventEncoder: Encoder[DeviceUpdateEvent] = Encoder.instance { x =>
+    x match {
+      case e: DeviceUpdateAvailable => deviceUpdateAvailableEncoder(e).mapObject(_.add("eventType", "DeviceUpdateAvailable".asJson))
```

Comment:
Any ideas on how to do this nicely?

Reply:
Guess this is only marginally better, but what do you think about:
```
def encodeDeviceUpdateEvent[T <: DeviceUpdateEvent](e : T)(implicit enc: Encoder[T]): Json =
    e.asJson.mapObject(_.add("eventType", e.getClass.getSimpleName.asJson))

  implicit val deviceUpdateEventEncoder: Encoder[DeviceUpdateEvent] = Encoder.instance {
    case e: DeviceUpdateAvailable => encodeDeviceUpdateEvent(e)
    case e: DeviceUpdateCanceled => encodeDeviceUpdateEvent(e)
    case e: DeviceUpdateCompleted => encodeDeviceUpdateEvent(e)
  }
```

Reply:
As it is a sealed hierarchy, `io.circe.generic.semiauto.deriveEncoder[DeviceUpdateEvent]` will should work. It will derive as encoder for coproduct in this case and add a discriminator.